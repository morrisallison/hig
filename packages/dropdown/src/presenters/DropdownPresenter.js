import React from "react";
import PropTypes from "prop-types";

import InputPresenter from "./InputPresenter";
import MenuPresenter from "./MenuPresenter";
import renderWrapper from "./WrapperPresenter";
import renderOptions from "./renderOptions";

/** @typedef {import("./renderOptions").OptionMeta} OptionMeta */
/** @typedef {import("downshift").ControllerStateAndHelpers} DownshiftHelpers */

function renderInput(props) {
  const {
    id,
    toggleMenu,
    getInputProps,
    label,
    instructions,
    placeholder,
    disabled,
    required,
    onBlur,
    onFocus
  } = props;

  const inputProps = getInputProps({
    id,
    label,
    instructions,
    placeholder,
    disabled,
    required,
    onBlur,
    onFocus,
    onClick: toggleMenu
  });

  return <InputPresenter key="input" {...inputProps} />;
}

function renderMenuWithOptions(props) {
  const {
    isOpen,
    getItemProps,
    getMenuProps,
    highlightedIndex,
    selectedItem,
    selectedItems,
    multiple,
    options,
    formatOption
  } = props;
  const menuProps = getMenuProps({ isOpen });
  const children = renderOptions({
    multiple,
    options,
    formatOption,
    getItemProps,
    highlightedIndex,
    selectedItem,
    selectedItems
  });

  return (
    <MenuPresenter key="menu" {...menuProps}>
      {children}
    </MenuPresenter>
  );
}

/**
 * @param {DownshiftHelpers} downshift
 * @returns {JSX.Element}
 */
export default function DropdownPresenter({ disabled }) {
  /**
   * The `Wrapper` presenter is used as a function to avoid having to use Downshift's `getRootProps`
   * @see https://github.com/paypal/downshift#getrootprops
   */
  return renderWrapper({
    disabled,
    children: [renderInput(), renderMenuWithOptions()]
  });
}

DropdownPresenter.propTypes = {
  /**
   * HTML ID attribute
   */
  id: PropTypes.string,
  /**
   * Text describing what the field represents
   */
  label: PropTypes.string,
  /**
   * Instructional text for the field
   */
  instructions: PropTypes.string,
  /**
   * Placeholder text to render when an option has not been selected
   */
  placeholder: PropTypes.string,
  /**
   * Enables multiple selection
   */
  multiple: PropTypes.bool,
  /**
   * Prevents user actions on the field
   */
  disabled: PropTypes.bool,
  /**
   * Text describing why the field is required
   */
  required: PropTypes.string,
  /**
   * An array of objects to choose from
   */
  value: PropTypes.oneOfType([PropTypes.any, PropTypes.arrayOf(PropTypes.any)]),
  /**
   * An array of unique values of any type except `undefined`
   */
  options: PropTypes.arrayOf(PropTypes.any),
  /**
   * Called when the text field is blurred
   */
  onBlur: PropTypes.func,
  /**
   * Called when the text field is focused
   */
  onFocus: PropTypes.func,
  /**
   * Used to format options into human readable strings
   */
  formatOption: PropTypes.func,
  /**
   * Calls the provided callback when the user presses a key while the dropdown has focus
   */
  // onKeydown: PropTypes.func

  /** @see Downshift */
  toggleMenu: PropTypes.func,
  /** @see Downshift */
  getInputProps: PropTypes.func,
  /** @see Downshift */
  isOpen: PropTypes.func,
  /** @see Downshift */
  getItemProps: PropTypes.func,
  /** @see Downshift */
  getMenuProps: PropTypes.func,
  /** @see Downshift */
  highlightedIndex: PropTypes.number,
  /** @see Downshift */
  selectedItem: PropTypes.any,
  /** @see Downshift */
  selectedItems: PropTypes.arrayOf(PropTypes.any)
};

DropdownPresenter.defaultProps = {
  getInputProps: () => ({}),
  getMenuProps: () => ({})
};
