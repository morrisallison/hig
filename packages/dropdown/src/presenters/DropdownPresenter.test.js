import renderer from "react-test-renderer";
import React from "react";

import DropdownPresenter from "./DropdownPresenter";

describe("DropdownPresenter", () => {
  const i18n = {
    foo: "Foo",
    bar: "Bar",
    baz: "Baz"
  };

  [
    {
      desc: "renders without props",
      props: {}
    },
    {
      desc: "renders with a label and an ID",
      props: {
        id: "whoaBuddy",
        label: "HIG Themes"
      }
    },
    {
      desc: "renders with custom option formatting",
      props: {
        options: Object.keys(i18n).sort(),
        formatOption: key => i18n[key]
      }
    },
    {
      desc: "renders non-undefined values as options",
      props: {
        options: ["foo", 1, true, false, null, [], {}, new Map()]
      }
    }
  ].forEach(({ desc, props }) => {
    it(desc, () => {
      const tree = renderer.create(<DropdownPresenter {...props} />).toJSON();

      expect(tree).toMatchSnapshot();
    });
  });
});
