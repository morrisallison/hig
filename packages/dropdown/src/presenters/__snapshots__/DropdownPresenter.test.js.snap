// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DropdownPresenter renders non-undefined values as options 1`] = `
<div
  className="hig__dropdown"
>
  <div
    className="hig__dropdown__input-wrapper"
  >
    <div
      className="hig__text-field-v1 hig__text-field-v1--button"
    >
      <div
        className="hig__text-field-v1__content"
      >
        <div
          className="hig__text-field-v1__input-wrapper"
        >
          <div
            className="hig__text-field-v1__input-row"
          >
            <input
              className="hig__text-field-v1__input hig__text-field-v1__input--no-value"
              id="text-field-65308"
              onBlur={[Function]}
              onFocus={[Function]}
              readOnly={true}
              type="button"
            />
          </div>
        </div>
      </div>
    </div>
    <span
      className="hig__dropdown__input-caret"
    >
      <div
        className="hig__icon hig__icon--24-size"
        dangerouslySetInnerHTML={
          Object {
            "__html": "<svg width=\\"10\\" height=\\"6\\" viewBox=\\"0 0 10 6\\" xmlns=\\"http://www.w3.org/2000/svg\\" role=\\"img\\" title=\\"caret-24\\"><polyline transform=\\"rotate(45 9.828 3)\\" points=\\"8 2 8 8 2 8\\" stroke=\\"#2E4258\\" stroke-width=\\"1.15\\" fill=\\"none\\" fill-rule=\\"evenodd\\" stroke-linecap=\\"round\\" stroke-linejoin=\\"round\\"/></svg>",
          }
        }
      />
    </span>
  </div>
  <div
    className="hig__dropdown-v1__menu"
  >
    <div
      className="hig__dropdown-option"
      index={0}
      item="foo"
    >
      <span
        className="hig__dropdown-option__label"
      >
        foo
      </span>
    </div>
    <div
      className="hig__dropdown-option"
      index={1}
      item={1}
    >
      <span
        className="hig__dropdown-option__label"
      >
        1
      </span>
    </div>
    <div
      className="hig__dropdown-option"
      index={2}
      item={true}
    >
      <span
        className="hig__dropdown-option__label"
      >
        true
      </span>
    </div>
    <div
      className="hig__dropdown-option"
      index={3}
      item={false}
    >
      <span
        className="hig__dropdown-option__label"
      >
        false
      </span>
    </div>
    <div
      className="hig__dropdown-option"
      index={4}
      item={null}
    >
      <span
        className="hig__dropdown-option__label"
      >
        null
      </span>
    </div>
    <div
      className="hig__dropdown-option"
      index={5}
      item={Array []}
    >
      <span
        className="hig__dropdown-option__label"
      >
        
      </span>
    </div>
    <div
      className="hig__dropdown-option"
      index={6}
      item={Object {}}
    >
      <span
        className="hig__dropdown-option__label"
      >
        [object Object]
      </span>
    </div>
    <div
      className="hig__dropdown-option"
      index={7}
      item={Map {}}
    >
      <span
        className="hig__dropdown-option__label"
      >
        [object Map]
      </span>
    </div>
  </div>
</div>
`;

exports[`DropdownPresenter renders with a label and an ID 1`] = `
<div
  className="hig__dropdown"
>
  <div
    className="hig__dropdown__input-wrapper"
  >
    <div
      className="hig__text-field-v1 hig__text-field-v1--button"
    >
      <div
        className="hig__text-field-v1__content"
      >
        <div
          className="hig__text-field-v1__input-wrapper"
        >
          <div
            className="hig__text-field-v1__input-row"
          >
            <input
              className="hig__text-field-v1__input hig__text-field-v1__input--no-value"
              id="text-field-65308"
              onBlur={[Function]}
              onFocus={[Function]}
              readOnly={true}
              type="button"
            />
          </div>
        </div>
      </div>
    </div>
    <span
      className="hig__dropdown__input-caret"
    >
      <div
        className="hig__icon hig__icon--24-size"
        dangerouslySetInnerHTML={
          Object {
            "__html": "<svg width=\\"10\\" height=\\"6\\" viewBox=\\"0 0 10 6\\" xmlns=\\"http://www.w3.org/2000/svg\\" role=\\"img\\" title=\\"caret-24\\"><polyline transform=\\"rotate(45 9.828 3)\\" points=\\"8 2 8 8 2 8\\" stroke=\\"#2E4258\\" stroke-width=\\"1.15\\" fill=\\"none\\" fill-rule=\\"evenodd\\" stroke-linecap=\\"round\\" stroke-linejoin=\\"round\\"/></svg>",
          }
        }
      />
    </span>
  </div>
  <div
    className="hig__dropdown-v1__menu"
  />
</div>
`;

exports[`DropdownPresenter renders with custom option formatting 1`] = `
<div
  className="hig__dropdown"
>
  <div
    className="hig__dropdown__input-wrapper"
  >
    <div
      className="hig__text-field-v1 hig__text-field-v1--button"
    >
      <div
        className="hig__text-field-v1__content"
      >
        <div
          className="hig__text-field-v1__input-wrapper"
        >
          <div
            className="hig__text-field-v1__input-row"
          >
            <input
              className="hig__text-field-v1__input hig__text-field-v1__input--no-value"
              id="text-field-65308"
              onBlur={[Function]}
              onFocus={[Function]}
              readOnly={true}
              type="button"
            />
          </div>
        </div>
      </div>
    </div>
    <span
      className="hig__dropdown__input-caret"
    >
      <div
        className="hig__icon hig__icon--24-size"
        dangerouslySetInnerHTML={
          Object {
            "__html": "<svg width=\\"10\\" height=\\"6\\" viewBox=\\"0 0 10 6\\" xmlns=\\"http://www.w3.org/2000/svg\\" role=\\"img\\" title=\\"caret-24\\"><polyline transform=\\"rotate(45 9.828 3)\\" points=\\"8 2 8 8 2 8\\" stroke=\\"#2E4258\\" stroke-width=\\"1.15\\" fill=\\"none\\" fill-rule=\\"evenodd\\" stroke-linecap=\\"round\\" stroke-linejoin=\\"round\\"/></svg>",
          }
        }
      />
    </span>
  </div>
  <div
    className="hig__dropdown-v1__menu"
  >
    <div
      className="hig__dropdown-option"
      index={0}
      item="bar"
    >
      <span
        className="hig__dropdown-option__label"
      >
        Bar
      </span>
    </div>
    <div
      className="hig__dropdown-option"
      index={1}
      item="baz"
    >
      <span
        className="hig__dropdown-option__label"
      >
        Baz
      </span>
    </div>
    <div
      className="hig__dropdown-option"
      index={2}
      item="foo"
    >
      <span
        className="hig__dropdown-option__label"
      >
        Foo
      </span>
    </div>
  </div>
</div>
`;

exports[`DropdownPresenter renders without props 1`] = `
<div
  className="hig__dropdown"
>
  <div
    className="hig__dropdown__input-wrapper"
  >
    <div
      className="hig__text-field-v1 hig__text-field-v1--button"
    >
      <div
        className="hig__text-field-v1__content"
      >
        <div
          className="hig__text-field-v1__input-wrapper"
        >
          <div
            className="hig__text-field-v1__input-row"
          >
            <input
              className="hig__text-field-v1__input hig__text-field-v1__input--no-value"
              id="text-field-65308"
              onBlur={[Function]}
              onFocus={[Function]}
              readOnly={true}
              type="button"
            />
          </div>
        </div>
      </div>
    </div>
    <span
      className="hig__dropdown__input-caret"
    >
      <div
        className="hig__icon hig__icon--24-size"
        dangerouslySetInnerHTML={
          Object {
            "__html": "<svg width=\\"10\\" height=\\"6\\" viewBox=\\"0 0 10 6\\" xmlns=\\"http://www.w3.org/2000/svg\\" role=\\"img\\" title=\\"caret-24\\"><polyline transform=\\"rotate(45 9.828 3)\\" points=\\"8 2 8 8 2 8\\" stroke=\\"#2E4258\\" stroke-width=\\"1.15\\" fill=\\"none\\" fill-rule=\\"evenodd\\" stroke-linecap=\\"round\\" stroke-linejoin=\\"round\\"/></svg>",
          }
        }
      />
    </span>
  </div>
  <div
    className="hig__dropdown-v1__menu"
  />
</div>
`;
